feat: Add support for custom modules.

71 opened the pull request:
#### Description
Add custom modules to starship:

```toml
prompt_order = [
  # snip...
  "custom.foo",
  "custom.bar",
  # snip...
]

[custom.foo]
command = "echo foo"  # shows output of command
files = ["foo"]       # can specify filters
extensions = ["py"]   # can specify filters (cont)
prefix = " transcending "

[custom.bar]
command = "echo I will never get echo'd"
when = "false"  # if no filter matches, fall back to executing 'when'
```

#### Motivation and Context
I also want to have custom commands, especially for personal or company-specific commands and environments. This allows any sort of section to be added.

Closes #664, Closes #790

#### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### How Has This Been Tested?
I tested multiple configurations in the terminal, experimenting with various settings.

- [ ] I have tested using **MacOS**
- [X] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:
- [X] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

71 commented:
For now I'm mostly looking for feedback and possible suggestions, but it works well on my end so you can probably try it out and merge it if it's fine for you.

matchai commented on src/print.rs:
```diff
@@ -132,12 +132,45 @@ pub fn explain(args: ArgMatches) {
+                        module,
+                    ),
+                    None => log::debug!(
+                        "Expected prompt_order to contain a custom value, but none was provided. Instead received {}",
```
I wasn't totally clear on what the log message was saying here. Perhaps something along these lines should be used:
`prompt_order contains custom module "{}", but no configuration was provided. Configuration for the following modules were provided: {}`

71 replied:
Once again, good description! Thank you.

71 commented on docs/config/README.md:
```diff
@@ -314,10 +314,10 @@ This does not suppress conda's own prompt modifier, you may want to run `conda c
-| `symbol`            | `"C "`         | The symbol used before the environment name.                                                                                                                                                                |
 | `style`             | `"bold green"` | The style for the module.                                                                                                                                                                                   |
-| `disabled`          | `false`        | Disables the `conda` module.                                                                                                                                                                                |
+| `symbol`            | `"C "`         | The symbol used before the environment name.                                                                                                                                                                |
```
I did change that on accident, and reverted it.

71 commented on docs/config/README.md:
````diff
@@ -1181,3 +1181,53 @@ The module will be shown if any of the following conditions are met:
+```toml
+# ~/.config/starship.toml
+
+prompt_order = [
````
I wasn't certain about this either, but since you're okay with that, great! I'm enabling `toml`'s "preserve_order" feature for that, so that we know the order at which the custom modules were defined.

71 replied:
There are several ways to do this. Should I add an additional (magic-ish) "custom" module that adds all custom modules in the order they were defined and add it to the `prompt_order` (and cie), or should I handle that more directly in the printer?

matchai replied:
I think the former sounds like a clearer solution if it's not too difficult to do.
It would possibly allow users to define the location of "custom" in `prompt_order` as a group, if they choose to do so.

71 replied:
Here's what I did:
- Create a special "custom" module
  * It is special because it actually returns a vector of modules, rather than a list of modules.
  * It is therefore handled directly in `print.rs`.
- Add it to the default prompt order before `line_break`.
- When it gets loaded, it adds all custom modules **that are not explicitly added in `prompt_order`** in the order they were defined in the modules to print.

71 commented on src/modules/custom.rs:
```diff
@@ -0,0 +1,155 @@
+            is_match = exec_when(when);
+        }
+
+        if !is_match {
```
`is_match` might be stay `false` right above, though, if `exec_when` returns `false`.

matchai replied:
👍

71 commented:
(Sorry for force-pushing so much; I just think it's cleaner to have a single commit in that PR).

`custom` is now right before `line_break` in the default prompt order. It displays all custom modules that aren't already specified with the `custom.foo` syntax in the order they were defined.

I also added some tests and changed the docs a bit.

matchai commented:
I have no clue what could be causing the failures here... 🤔
Maybe some @starship/astronauts can help us troubleshoot?

chipbuster commented:
@matchai Do you know if there's a way to set the RUST_LOG variable for these tests? I see a couple possible failure points within `exec_command()` and `shell_command()`, and we actually have a few trace statements there, but we aren't logging at the moment so it's tricky to tell.

(The tests all pass on my system, so I'm afraid I'm not much use there)

matchai commented:
Unfortunately, the logger is not initialized in the test binary.
You would need to add the logger initialization on a branch, set the environment variable, and run the tests with `--no-capture` to see any logs.

71 commented on src/modules/custom.rs:
```diff
@@ -0,0 +1,202 @@
+}
+
+/// Return the invoking shell (using STARSHIP_SHELL), or "sh"
+fn shell() -> String {
```
Good idea! I just added this feature to the fork.

71 commented:
Thanks for the idea @diminishedprime. Plus, the feature can be used to explicitly use `/bin/sh` in the tests, which fixes our "tests don't pass on VMs" problem. Yay!

71 commented:
Windows was a _little_ hard to get to work, but I think I've got something good now. Can the PR be reviewed again?

diminishedprime commented on src/modules/custom.rs:
```diff
@@ -0,0 +1,288 @@
+        module.get_prefix().set_value(prefix);
+    }
+    if let Some(suffix) = config.suffix {
+        module.get_prefix().set_value(suffix);
```
```suggestion
        module.get_suffix().set_value(suffix);
```

71 replied:
Good catch!

71 commented:
@andytom Thanks for the suggestion. I did notice those utils at the very beginning, but decided not to use them because the whole feature relies on running actual commands. I thought that mocking the one part that is the most likely to break would be risky, and instead went with reasonably common utilities in the commands I run.

I also wanted to pass custom `stdin`, which isn't possible with the current tests. `utils` could be updated for my needs, but it seems like my requirements were quite different from what other modules typically need, so I went with some custom, specialized code.
